### PR TITLE
Refactor paging from recursive mapping to linear mapping

### PR DESCRIPTION
**Pull request by wangrunji0408**
## Motivation
The current page table use recursive mapping to access itself and do not have a linear mapping to all physical memory.

This mechanism follows BlogOS. It worked well at the beginning.

But we found several limitations as our OS grows:
1. It's a really complicated concept that hard to understand for beginners. 
2. For some architectures, recursive mapping can not be implemented normally. For example, RISC-V and MIPS.
3. It's not easy to access page table and data of other processes. So forking process and coping the memory can not be implemented both simply and efficiently.
4. All access to an inactive page table needs to modify the active one, which causes unnecessary data race on multi-thread and multi-core.
5. It discourages us to access physical memory directly, which is unfriendly to driver and DMA.

Almost everyone agrees that linear mapping is a better choice.

## Changes
1. Merge `InactivePageTable` and `PageTable` trait.
2. Add `phys_to_virt` function.
3. Remove all DMA mappings in drivers.
4. ...

## Progress

|Arch|Status|Note|
|----|----|----|
| x86_64 |✅|  |
| RISC-V |✅|  |
| AArch64 |✅|  |
| MIPS |✅| |

### CODE REVIEW

**Review by miskcoo — COMMENTED** [review]
It looks good for MIPS.

File: kernel/src/drivers/block/virtio_blk.rs [hunk 1]
```diff
@@ -127,9 +125,10 @@ impl Driver for VirtIOBlkDriver {
 
     fn read_block(&self, block_id: usize, buf: &mut [u8]) -> bool {
         let mut driver = self.0.lock();
-        // ensure header page is mapped
-        active_table().map_if_not_exists(driver.header as usize, driver.header as usize);
+<<<<<<< HEAD
```
Comment (Harry-Chen) [inline kernel/src/drivers/block/virtio_blk.rs]: Why is here merge conflicts?

**Review by Harry-Chen — CHANGES_REQUESTED** [review]
MIPS part looks good to me, but RISC-V part seems not done yet

**Comment by jiegec** [comment]
Still one problem remains: #[repr(align(64)]] does not work with rv64.

**Comment by jiegec** [comment]
RV32+64 is ready now.

**Review by Harry-Chen — APPROVED** [review]

**Review by jiegec — APPROVED** [review]

**Comment by Harry-Chen** [comment]
I believe it is now feasible to merge these commits

**Comment by wangrunji0408** [comment]
Then aarch64 is broken. 😂
@equation314 Could you help fix linear mapping on aarch64?

**Comment by equation314** [comment]
OK, I will try to work in it this week...

**Pull request merged**